### PR TITLE
402 Surrogate model implementation with multiple age groups and dampings 

### LINKED ISSUE #402: Neural net for multiple age groups
We want to create a neural network that predicts the developement of the epidemic situation for an ODE model with a population that is  subdivided into six age groups.
This increased model complexity, since more information has to be processed. 
Wheras the contact behavior was described  by a number in the model with one age group, the contact behavior now is represented by a matrix that contains the contact frequencies from one age group with another.

Further, we will not consider any dampings, i.e., no change contact patterns.

### LINKED ISSUE #403: Predict the development of the pandemic after one damping at a fixed day
We want to create a neural network that predicts the developement of the epidemic situation for an ODE model that includes changes of contact patterns among different age groups. 

In a pandemic, the government can enact measures that reduce the contact between citizen. Since the diseases are transmitted via human contact, the contact reduction influences the progression of the disease. 

The models describe the development of the pandemic after one reduction of contacts (damping) at a fixed day. 
The input for the model now contains the contact matrix additionally to the distribution of the population among the compartments. 
Three types of models can be trained: a multilayer percepton, a convolutional neural net and a long short-term memory model.

### PR DESCRIPTION
**Pull request by HenrZu**
## Merge Request - GuideLine Checklist 

**Guideline** to check code before resolve WIP and approval, respectively.
As many checkboxes as possible should be ticked.

### Checks by code author:
Always to be checked:
* [x] There is at least one issue associated with the pull request.
* [x] The branch follows the naming conventions as defined in the [git workflow](git-workflow).
* [x] New code adheres with the [coding guidelines](coding-guidelines)
* [x] Make sure that the [pre-commit linting/style checks pass](https://github.com/DLR-SC/memilio/wiki).

If functions were changed or functionality was added:
* [x] Tests for new functionality has been added
* [x] A local test was succesful

If new functionality was added:
* [x] There is appropriate **documentation** of your work. (use doxygen style comments)

If new third party software is used:
* [x] Did you pay attention to its license? Please remember to add it to the wiki after successful merging.

If new mathematical methods or epidemiological terms are used:
* [x] Are new methods referenced? Did you provide further documentation? Has the glossary been updated? 

The following questions are addressed in the documentation if need be: 
* [x] Developers (what did you do?, how can it be maintained?)
* [x] For users (how to use your work?)
* [x] For admins (how to install and configure your work?)

* For documentation: Please write or update the Readme in the current working directory!

### Checks by code reviewer(s):
* [x] Is the code clean of development artifacts e.g., unnecessary comments, prints, ...
* [x] The ticket goals for each associated issue are reached or problems are clearly addressed (i.e., a new issue was introduced).
* [x] There are appropriate **unit tests** and they pass.
* [x] The git history is clean and linearized for the merge request. All reviewers should squash commits and write a simple and meaningful commit message.
* [x] Coverage report for new code is acceptable. 

Closes #402, Closes #403

### CODE REVIEW

File: pycode/memilio-surrogatemodel/README.md [hunk 1]
```diff
@@ -24,6 +24,7 @@ The package currently provides the following modules:
 - `models`: models for different specific tasks
    Currently we have the following models: 
    - `ode_secir_simple`: A simple model allowing for asymptomatic as well as symptomatic infection not stratified by age groups.
+   - `ode_secir_groups`: A model allowing for asymptomatic as well as symptomatic infection stratified by age groups and one damping.
```
Comment (jubicker) [inline pycode/memilio-surrogatemodel/README.md:27]:
```suggestion
   - `ode_secir_groups`: A model allowing for asymptomatic as well as symptomatic infection stratified by age groups and including one damping.
```

File: pycode/memilio-surrogatemodel/memilio/surrogatemodel/ode_secir_groups/data_generation.py [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+from memilio.simulation.secir import (AgeGroup, Index_InfectionState,
+                                      InfectionState, Model,
+                                      interpolate_simulation_result, simulate)
+
+
+def run_secir_groups_simulation(days, damping_day, populations):
+    """! Uses an ODE SECIR model allowing for asymptomatic infection with 6 different age groups. The model is not stratified by region. 
+    Virus-specific parameters are fixed and initial number of persons in the particular infection states are chosen randomly from defined ranges.
+    @param Days Describes how many days we simulate within a single run.
+    @return List containing the populations in each compartment used to initialize the run.
```
Comment (jubicker) [inline pycode/memilio-surrogatemodel/memilio/surrogatemodel/ode_secir_groups/data_generation.py:42]: What about the description of the parameters damping_day and populations?

File: pycode/memilio-surrogatemodel/memilio/surrogatemodel/ode_secir_groups/data_generation.py [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+        model.parameters.TransmissionProbabilityOnContact[AgeGroup(i)] = 0.1
+        model.parameters.RecoveredPerInfectedNoSymptoms[AgeGroup(i)] = 0.09
+        model.parameters.RiskOfInfectionFromSymptomatic[AgeGroup(i)] = 0.25
+        model.parameters.SeverePerInfectedSymptoms[AgeGroup(i)] = 0.2
+        model.parameters.CriticalPerSevere[AgeGroup(i)] = 0.25
+        model.parameters.DeathsPerCritical[AgeGroup(i)] = 0.3
+        # twice the value of RiskOfInfectionFromSymptomatic
+        model.parameters.MaxRiskOfInfectionFromSymptomatic[AgeGroup(i)] = 0.5
+
+    model.parameters.StartDay = (
```
Comment (jubicker) [inline pycode/memilio-surrogatemodel/memilio/surrogatemodel/ode_secir_groups/data_generation.py:102]: Maybe add a comment why StartDay is set like this

File: pycode/memilio-surrogatemodel/memilio/surrogatemodel/ode_secir_groups/data_generation.py [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+
+    # The number of days is the same as the sum of input and label width.
+    # Since the first day of the input is day 0, we still need to subtract 1.
+    days = input_width + label_width - 1
+
+    # Load population data
+    population = get_population(path_population)
+
+    # show progess in terminal for longer runs
+    # Due to the random structure, theres currently no need to shuffle the data
```
Comment (jubicker) [inline pycode/memilio-surrogatemodel/memilio/surrogatemodel/ode_secir_groups/data_generation.py:173]:
```suggestion
    # Due to the random structure, there's currently no need to shuffle the data
```

File: pycode/memilio-surrogatemodel/memilio/surrogatemodel/ode_secir_groups/data_generation.py [hunk 5]
```diff
@@ -0,0 +1,302 @@
...
+    for _ in range(0, num_runs):
+
+        # Generate a random damping day
+        damping_day = random.randrange(
+            input_width, input_width+label_width)
+
+        data_run, damped_contact_matrix = run_secir_groups_simulation(
+            days, damping_day, population[random.randint(0, len(population) - 1)])
+        data['inputs'].append(data_run[:input_width])
+        data['labels'].append(data_run[input_width:])
```
Comment (jubicker) [inline pycode/memilio-surrogatemodel/memilio/surrogatemodel/ode_secir_groups/data_generation.py:184]: Above you commented that the first day of the input is day 0, but data_run is returning the values from day 1 on. So you are not using day 0 as an input. Instead your first input is day 1 of the simulation?

File: pycode/memilio-surrogatemodel/memilio/surrogatemodel/ode_secir_groups/data_generation.py [hunk 6]
```diff
@@ -0,0 +1,302 @@
...
+        data['contact_matrix'].append(np.array(damped_contact_matrix))
+        data['damping_day'].append(damping_day)
+        bar.next()
+    bar.finish()
+
+    if normalize:
+        # logarithmic normalization
+        transformer = FunctionTransformer(np.log1p, validate=True)
+        inputs = np.asarray(data['inputs']).transpose(2, 0, 1).reshape(48, -1)
```
Comment (jubicker) [inline pycode/memilio-surrogatemodel/memilio/surrogatemodel/ode_secir_groups/data_generation.py:193]: Maybe add a comment why you are reshapung the inputs like this

File: pycode/memilio-surrogatemodel/memilio/surrogatemodel/ode_secir_groups/data_generation.py [hunk 7]
```diff
@@ -0,0 +1,302 @@
...
+        scaled_inputs = transformer.transform(inputs)
+        scaled_inputs = scaled_inputs.transpose().reshape(num_runs, input_width, 48)
```
Comment (jubicker) [inline pycode/memilio-surrogatemodel/memilio/surrogatemodel/ode_secir_groups/data_generation.py:195]: input dimension is not input_width x num_compartments x num_age_groups any more but is now input_width x 48

File: pycode/memilio-surrogatemodel/memilio/surrogatemodel/ode_secir_groups/data_generation.py [hunk 8]
```diff
@@ -0,0 +1,302 @@
...
+        scaled_inputs_list = scaled_inputs.tolist()
+
+        labels = np.asarray(data['labels']).transpose(2, 0, 1).reshape(48, -1)
+        scaled_labels = transformer.transform(labels)
+        scaled_labels = scaled_labels.transpose().reshape(num_runs, label_width, 48)
```
Comment (jubicker) [inline pycode/memilio-surrogatemodel/memilio/surrogatemodel/ode_secir_groups/data_generation.py:200]: same as above with input shapes

File: pycode/memilio-surrogatemodel/memilio/surrogatemodel/ode_secir_groups/data_generation.py [hunk 9]
```diff
@@ -0,0 +1,302 @@
...
+        scaled_labels_list = scaled_labels.tolist()
+
+        # cast dfs to tensors
+        data['inputs'] = tf.stack(scaled_inputs_list)
```
Comment (jubicker) [inline pycode/memilio-surrogatemodel/memilio/surrogatemodel/ode_secir_groups/data_generation.py:204]: if you do not normalize you don't need to cast to tensors?

File: pycode/memilio-surrogatemodel/memilio/surrogatemodel/ode_secir_groups/data_generation.py [hunk 10]
```diff
@@ -0,0 +1,302 @@
...
+    @param path Path to the dataset containing the population data
+    """
+
+    with open(path) as f:
+        data = json.load(f)
+    population = []
+    for data_entry in data:
+        population_county = []
+        population_county.append(
+            data_entry['<3 years'] + data_entry['3-5 years'] / 2)
```
Comment (jubicker) [inline pycode/memilio-surrogatemodel/memilio/surrogatemodel/ode_secir_groups/data_generation.py:271]: Why is '3-5 years' divided by 2 and where does the rest of the '3-5' population go?

File: pycode/memilio-surrogatemodel/memilio/surrogatemodel/ode_secir_groups/model.py [hunk 11]
```diff
@@ -0,0 +1,504 @@
...
+
+    @param inputs test inputs for model prediction. 
+    @param labels test labels. 
+    @param modeltype type of model. Can be 'classic' or 'timeseries'
+    @param model trained model. 
+    @param plot_col string name of compartment to be plotted. 
+    @param max_subplots Number of the simulation runs to be plotted and compared against. 
+    """
+    if modeltype == 'classic':
+        input_width = int((inputs.shape[1] - 37) / 48)
```
Comment (jubicker) [inline pycode/memilio-surrogatemodel/memilio/surrogatemodel/ode_secir_groups/model.py:47]: Maybe add a comment where the 37 and 48 comes from

File: pycode/memilio-surrogatemodel/memilio/surrogatemodel/ode_secir_groups/model.py [hunk 12]
```diff
@@ -0,0 +1,504 @@
...
+    for n in range(max_n):
+        plt.subplot(max_n, 1, n+1)
+        plt.ylabel(plot_compartment)
+
+        input_array = inputs[n].numpy()
+        label_array = labels[n].numpy()
+
+        if modeltype == 'classic':
+            input_plot = input_array[:(input_width*6*8)]
+            input_plot = input_plot.reshape(5, 48)
```
Comment (jubicker) [inline pycode/memilio-surrogatemodel/memilio/surrogatemodel/ode_secir_groups/model.py:73]: If you need to use hard coded numbers please add a comment what they mean

File: pycode/memilio-surrogatemodel/memilio/surrogatemodel/ode_secir_groups/model.py [hunk 13]
```diff
@@ -0,0 +1,504 @@
...
+    if split_train + split_valid + split_test != 1:
+        ValueError("summed Split ratios not equal 1! Please adjust the values")
+
+    n = contact_matrices.shape[0]
+    n_train = int(n * split_train)
+    n_valid = int(n * split_valid)
+    n_test = int(n * split_test)
+
+    if n_train + n_valid + n_test != n:
+        n_test = n - n_train - n_valid
```
Comment (jubicker) [inline pycode/memilio-surrogatemodel/memilio/surrogatemodel/ode_secir_groups/model.py:415]: You could set n_test directly to n-n_train-n_valid

File: pycode/memilio-surrogatemodel/memilio/surrogatemodel/ode_secir_groups/model.py [hunk 14]
```diff
@@ -0,0 +1,504 @@
...
+    if split_train + split_valid + split_test != 1:
+        ValueError("summed Split ratios not equal 1! Please adjust the values")
+    damping_days = np.asarray(damping_days)
+    n = damping_days.shape[0]
+    n_train = int(n * split_train)
+    n_valid = int(n * split_valid)
+    n_test = int(n * split_test)
+
+    if n_train + n_valid + n_test != n:
+        n_test = n - n_train - n_valid
```
Comment (jubicker) [inline pycode/memilio-surrogatemodel/memilio/surrogatemodel/ode_secir_groups/model.py:447]: see above

File: pycode/memilio-surrogatemodel/memilio/surrogatemodel/ode_secir_groups/network_architectures.py [hunk 15]
```diff
@@ -0,0 +1,102 @@
...
+        tf.keras.layers.Reshape([1, -1]), ])
+    return model
+
+
+def mlp_multi_input_multi_output(label_width, num_age_groups=6):
+    """! Simple MLP Network which takes the compartments for multiple time steps as input and
+    returns the 8 compartments for all age groups for multiple time steps in the future. 
+
+    Reshaping adds an extra dimension to the output, so the shape of the output is 30x48.
+    This makes the shape comparable to that of the multi-output models.
```
Comment (jubicker) [inline pycode/memilio-surrogatemodel/memilio/surrogatemodel/ode_secir_groups/network_architectures.py:46]: Isn't the shape num_time_stepsx48 so only 30x48 if you want to predict 30 timesteps?

File: pycode/memilio-surrogatemodel/memilio/surrogatemodel_test/test_surrogatemodel_ode_secir_groups.py [hunk 16]
```diff
@@ -0,0 +1,266 @@
...
+        self.assertLessEqual(simulation_2[1].max(), 1)
+        self.assertGreaterEqual(simulation_2[1].max(), 0)
+
+        self.assertLessEqual(simulation_3[1].max(), 1)
+        self.assertGreaterEqual(simulation_3[1].max(), 0)
+
+    @patch('memilio.surrogatemodel.ode_secir_groups.data_generation.getMinimumMatrix',
+           return_value=0.1 * np.ones((6, 6)))
+    @patch('memilio.surrogatemodel.ode_secir_groups.data_generation.getBaselineMatrix',
+           return_value=0.6 * np.ones((6, 6)))
```
Comment (jubicker) [inline pycode/memilio-surrogatemodel/memilio/surrogatemodel_test/test_surrogatemodel_ode_secir_groups.py:91]: This is the same as above

File: pycode/memilio-surrogatemodel/memilio/surrogatemodel_test/test_surrogatemodel_ode_secir_groups.py [hunk 17]
```diff
@@ -0,0 +1,266 @@
...
+        data_generation.generate_data(num_runs, self.path, "", input_width,
+                                      label_width)
+        self.assertEqual(len(os.listdir(self.path)), 1)
+
+        self.assertEqual(os.listdir(self.path),
+                         ['data_secir_groups.pickle'])
+
+    def test_structures_networks(self):
+
+        label_width = 30
```
Comment (jubicker) [inline pycode/memilio-surrogatemodel/memilio/surrogatemodel_test/test_surrogatemodel_ode_secir_groups.py:150]: You don't use this variable

File: pycode/memilio-surrogatemodel/memilio/surrogatemodel/ode_secir_groups/network_architectures.py [hunk 18]
```diff
@@ -0,0 +1,102 @@
...
+        tf.keras.layers.Dense(units=32, activation='relu'),
+        tf.keras.layers.Dense(units=label_width*num_age_groups*8),
+        tf.keras.layers.Reshape([label_width, num_age_groups*8])
+    ])
+    return model
+
+
+def cnn_multi_input_multi_output(label_width, num_age_groups=6):
+    """! CNN Network which uses multiple time steps as input and returns the 8 compartments for
+    each age groups for multiple time step in the future.
```
Comment (mknaranja) [inline pycode/memilio-surrogatemodel/memilio/surrogatemodel/ode_secir_groups/network_architectures.py:63]:
```suggestion
    each age group for multiple time steps in the future.
```

File: pycode/memilio-surrogatemodel/memilio/surrogatemodel/ode_secir_groups/network_architectures.py [hunk 19]
```diff
@@ -0,0 +1,102 @@
...
+        tf.keras.layers.Dense(
+            label_width * num_age_groups * 8, activation='linear'),
+        tf.keras.layers.Reshape([label_width, 8 * num_age_groups])
+    ])
+    return model
+
+
+def lstm_multi_input_multi_output(label_width, num_age_groups=6):
+    """! LSTM Network which uses multiple time steps as input and returns the 8 compartments for
+    one single time step in the future.
```
Comment (jubicker) [inline pycode/memilio-surrogatemodel/memilio/surrogatemodel/ode_secir_groups/network_architectures.py:90]:
```suggestion
    multiple time steps in the future.
```

File: pycode/memilio-surrogatemodel/memilio/surrogatemodel/ode_secir_groups/model.py [hunk 20]
```diff
@@ -0,0 +1,505 @@
...
+
+            plt.plot(
+                np.arange(0, input_width),
+                mean_per_day_input,
+                label='Inputs', marker='.', zorder=-10)
+
+        elif modeltype == 'timeseries':
+            mean_per_day_input = []
+            for i in input_array:
+                x = i[plot_compartment_index: inputs.shape[1] - 37:8]
```
Comment (jubicker) [inline pycode/memilio-surrogatemodel/memilio/surrogatemodel/ode_secir_groups/model.py:92]: why -37? maybe add a comment to that

File: pycode/memilio-surrogatemodel/memilio/surrogatemodel_test/test_surrogatemodel_ode_secir_groups.py [hunk 21]
```diff
@@ -0,0 +1,265 @@
...
+
+        data_1 = data_generation.generate_data(
+            num_runs_1, self.path, "", input_width_1, label_width_1,
+            save_data=False)
+        self.assertEqual(len(data_1['inputs']), num_runs_1)
+        self.assertEqual(len(data_1['inputs'][0]), input_width_1)
+        self.assertEqual(len(data_1['inputs'][0][0]), 48)
+        self.assertEqual(len(data_1['labels']), num_runs_1)
+        self.assertEqual(len(data_1['labels'][0]), label_width_1)
+        self.assertEqual(len(data_1['labels'][0][0]), 48)
```
Comment (jubicker) [inline pycode/memilio-surrogatemodel/memilio/surrogatemodel_test/test_surrogatemodel_ode_secir_groups.py:115]: line 113-115 are doubled with lines 110-112

File: pycode/memilio-surrogatemodel/memilio/surrogatemodel_test/test_surrogatemodel_ode_secir_groups.py [hunk 22]
```diff
@@ -0,0 +1,265 @@
...
+
+        data_2 = data_generation.generate_data(
+            num_runs_2, self.path, "", input_width_2, label_width_2,
+            save_data=False)
+        self.assertEqual(len(data_2['inputs']), num_runs_2)
+        self.assertEqual(len(data_2['inputs'][0]), input_width_2)
+        self.assertEqual(len(data_2['inputs'][0][0]), 48)
+        self.assertEqual(len(data_2['labels']), num_runs_2)
+        self.assertEqual(len(data_2['labels'][0]), label_width_2)
+        self.assertEqual(len(data_2['labels'][0][0]), 48)
```
Comment (jubicker) [inline pycode/memilio-surrogatemodel/memilio/surrogatemodel_test/test_surrogatemodel_ode_secir_groups.py:125]: lines are doubled

**Review by jubicker — APPROVED** [review]

**Pull request merged**